### PR TITLE
Ignore unknown config file options; warn instead of error

### PR DESCRIPTION
**Pull request by sipa**
As reported by @satwo on IRC a few days ago, the current mechanism of treating unknown config file options as errors is problematic for options like `-rpcclienttimeout` which aren't defined for `bitcoind`.

A full solution would be to either make all binaries be aware of each other's options, or to permit config file options that only apply to specific binaries (`bitcoind`, `bitcoin-qt`, `bitcoin-cli`). Both of these seem too invasive to introduce for 0.17.

As a compromise, this PR makes it ignores those options, but still warn about it in the log file.

### CODE REVIEW

**Comment by ken2812221** [comment]
Related #13441

File: src/interfaces/node.cpp [hunk 1]
```diff
@@ -53,7 +53,7 @@ class NodeImpl : public Node
     {
         return gArgs.ParseParameters(argc, argv, error);
     }
-    bool readConfigFiles(std::string& error) override { return gArgs.ReadConfigFiles(error); }
+    bool readConfigFiles(std::string& error) override { return gArgs.ReadConfigFiles(error, true); }
```
Comment (MarcoFalke) [inline src/interfaces/node.cpp]: Please use named arguments for literal primitive types. (`/* ignore_invalid_keys */ true`)

**Comment by laanwj** [comment]
Tested ACK, we've discussed this on IRC and I think this is a good work-around for now.

Without:
```
$ src/bitcoind -rpcclienttimeout=1
Error parsing command line arguments: Invalid parameter -rpcclienttimeout
$ src/bitcoind -conf=/tmp/test.conf
Error reading configuration file: Invalid configuration value rpcclienttimeout
```

With:
```
$ src/bitcoind -rpcclienttimeout=1
Error parsing command line arguments: Invalid parameter -rpcclienttimeout
$ src/bitcoind -conf=/tmp/test.conf
[...]
```
(where `/tmp/test.conf` contains `rpcclienttimeout=1`)

It corrrectly rejects unknown direct arguments, but not those in the specified config file.

> A full solution would be to either make all binaries be aware of each other's options

I strongly dislike this solution, it destroys all semblance of decoupling, modularity.

> or to permit config file options that only apply to specific binaries (bitcoind, bitcoin-qt, bitcoin-cli). 

This would be better, command-specific config sections would improve readability of configuration as well.

**Comment by MarcoFalke** [comment]
utACK 5901ca411b420fe097e688425d624c7a8d9f1a30. Needs tests fixed/disabled.

**Comment by promag** [comment]
utACK 5901ca4.

How about hardcode the exceptions for now and keep the error behavior?

**Comment by laanwj** [comment]
> How about hardcode the exceptions for now and keep the error behavior?

No, let's not do that. This solution is good. The implementation of errors for unknown arguments was primarily introduced for command-line arguments anyhow—see #1044—at least I never predicted the implications for configuration options, but it's a separate concern.

**Comment by satwo** [comment]
Tested ACK 5901ca411b420fe097e688425d624c7a8d9f1a30.

File: src/util.cpp [hunk 2]
```diff
@@ -859,9 +859,13 @@ bool ArgsManager::ReadConfigStream(std::istream& stream, std::string& error, boo
...
         // Check that the arg is known
-        if (!IsArgKnown(strKey) && !ignore_invalid_keys) {
-            error = strprintf("Invalid configuration value %s", option.first.c_str());
-            return false;
+        if (!IsArgKnown(strKey)) {
+            if (!ignore_invalid_keys) {
+                error = strprintf("Invalid configuration value %s", option.first.c_str());
+                return false;
+            } else {
+                LogPrintf("Ignoring unknown configuration value %s\n", option.first);
```
Comment (achow101) [inline src/util.cpp]: This doesn't work because logging is not setup at this point in startup. The warning does not appear in the debug.log.

**Review by achow101 — COMMENTED** [review]
The test `feature_includeconf.py` has a test case for invalid options in the conf file. That will need to be removed or be updated.

File: src/util.cpp (hunk 2 above)
Comment (sipa) [reply src/util.cpp]: Hmm, I thought we cached early log messages and printed them afterwards. Any suggestions for how to deal with that?

Comment (sipa) [reply src/util.cpp]: Ah, #13088.

**Comment by achow101** [comment]
Right now, to get around this problem, we already have a bunch of hidden arguments. I think it would be reasonable, at least for 0.17, to just extend that list to include the bitcoin-cli arguments. Then later we can add more invasive changes to actually fix the problem.

**Comment by laanwj** [comment]
re-utACK 247d5740d29752c35861136a2fc561831c7e9832

**Comment by laanwj** [comment]
> I think it would be reasonable, at least for 0.17, to just extend that list to include the bitcoin-cli arguments.

No, I don't think that's reasonable. This solution is much better (I feel like I'm repeating myself...).

**Pull request merged**